### ISSUE TITLE
Crash after a few seconds On MacOS [M2]

### ISSUE THREAD

**Issue by SnorlaxAssist** [justDeeevin/NuhxBoard#8]
```
nuhxboard-aarch64-apple-darwin/nuhxboard ; exit;
The application panicked (crashed).
Message:  index out of bounds: the len is 0 but the index is 0
Location: src/nuhxboard.rs:498

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: PoisonError { .. }
Location: /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.10/src/platform_impl/macos/app_state.rs:387

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.


[Process completed]
```
Also a small square window appears with the blue background, and no Keyboard showed up while trying to load a keyboard.

**Comment by justDeeevin** [comment]
Judging by this error message, this is an issue with `winit`. Let me see if they have any issues for it.

I haven't gotten the opportunity to test the app on MacOS. It should work in theory, but I'm not surprised there's a problem. Sorry about this!

**Comment by justDeeevin** [comment]
Wait, no, i misread the message. The real error is in my code.

Do you know what version this is?

**Comment by justDeeevin** [comment]
I think i know what might be causing this, and i doubt it's MacOS specific.

Question 1: when specifically does the crash occur? When you try to open the load keyboard menu, or when you click on a keyboard in the list?

Question 2: do you have any keyboards downloaded? I admit, the README isn't very clear on this, but you do need to actually download or create some keyboard layouts to use them. My suggestion: clone ThoNohT/NohBoard and copy the `keyboards` directory to `~/.local/share/NuhxBoard/`. NohBoard ships with plenty of great examples.

I'm planning on downloading and unpacking those examples as part of the install step for the next release. First-time use is one of NuhxBoard's biggest flaws right now.

**Comment by SnorlaxAssist** [comment]
> Question 1: when specifically does the crash occur? When you try to open the load keyboard menu, or when you click on a keyboard in the list?

Yes, I click on load keyboard, then the drop down, once I click on global, the app crashes.
> Question 2: do you have any keyboards downloaded? I admit, the README isn't very clear on this, but you do need to actually download or create some keyboard layouts to use them. My suggestion: clone ThoNohT/NohBoard and copy the keyboards directory to ~/.local/share/NuhxBoard/. NohBoard ships with plenty of great examples.

No I do not have any keyboards downloaded, but when I made a new empty folder, this crash still happens, on a non empty folder, with non style files, it still crashes.

After I downloaded the keyboards as you suggested, the UI loaded, but none of the keystrokes are registering, and the mouse movements/inputs are stuck, I do have `Input Monitoring` enabled for the Nuhxboard app, still nothing.

<img width="799" alt="Screenshot 2024-03-20 at 12 12 40 PM" src="https://github.com/justDeeevin/NuhxBoard/assets/57375992/d6132148-f74f-47c6-9168-dc1cfbe74b03">

Also when I try load global from `Nohboard` with `default.style` nothing shows up.

<img width="292" alt="Screenshot 2024-03-20 at 12 09 17 PM" src="https://github.com/justDeeevin/NuhxBoard/assets/57375992/2bf35caa-13bb-44d5-885c-a1bd78840d4c">

I'm on MacOS Sonoma 14.2.1 (23C71)

**Comment by justDeeevin** [comment]
Alright, so it doesn't crash with the keyboards copied. That's a relief. I'll add a TODO to improve the behavior when there aren't any available keyboards.

About the frozen ui: in Windows and Linux under Wayland, my current method for listening for input doesn't work when the window itself has focus. That's usually no problem, since NuhxBoard is made to be a sort of background app (passively recorded as an overlay on game footage), so in its standard use case it isn't focused. Something similar is probably happening on MacOS. Try giving another window focus while NuhxBoard is still visible and test if the UI is working then.

About the whole `global` thing. That directory is specifically for globally-accessible **styles**. I'm _pretty sure_ that's how it is in NohBoard, too. I'll double-check. At any rate, it certainly shouldn't appear in the keyboard category list. I'll push a patch to fix that right away.

**Comment by SnorlaxAssist** [comment]
I just got another random crash after just clicking on the blank blue screen, this rarely happens.
```
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: src/nuhxboard.rs:330

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: PoisonError { .. }
Location: /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.10/src/platform_impl/macos/app_state.rs:387

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
```
> Try giving another window focus while NuhxBoard is still visible and test if the UI is working then.

It is still frozen. Weird part, once I focused the NuhxBoard window, Mouse Inputs seem to work, but no keyboard input. After switching windows back and forth, the mouse inputs stop working, and ends up freezing again.

**Comment by justDeeevin** [comment]
crash actually occurs here:
https://github.com/justDeeevin/NuhxBoard/blob/53e289cee7a29652423b50ece10bac4a25abb61e/src/nuhxboard.rs#L330

I'm pretty confident this is happening because it's receiving an orphaned release event (no corresponding button press), which isn't surprising given that you're having problems with receiving input overall.

> ends up freezing again

Do you mean the keys aren't lighting up when you press them, or the app isn't responding at all (not even opening the right-click menu)? If it's the latter, then something even more strange is going on here...

**Comment by SnorlaxAssist** [comment]
Nothing lights up, and the mouse input shows as it is pressed and the velocity is to a direction, that's the freeze.

**Comment by justDeeevin** [comment]
Woah it's been more than a year since this issue got submitted

anyway

I've finally gotten a MacOS VM running to test NuhxBoard with. With the fork of `rdev` that I'm now using, everything seems to function fine on MacOS sequoia. I obviously can't test an M2 without getting myself a macbook, but please try out compiling and running the latest version of this repo on your computer!

**Comment by justDeeevin** [comment]
I've gotten a lab rat (thanks @TheHaloDeveloper) to test this on an M3 chip and the app is still failing to receive input events. Reason is unclear as of yet, and it's especially confusing since I've confirmed that intel macOS works fine.
